docs/sphinx/cookbook/resource_aware_pool.rst

```diff
@@ -0,0 +1,150 @@
+Similarly, when we are executing on the host, this
+corresponds to a separate "stream of execution" and therefore a separate Camp host resource. 
+
+Typically, we deal with multiple Camp resources. This includes a single resource for the host and 
+one or more for the device, depending on how many (cuda, hip, etc.) streams we have in use. 
+While we can have multiple camp resources for the device (e.g. multiple cuda streams), 
+we can only have one resource for the host because the host only has one stream of execution. 
+Since we are dealing with Camp resources, we call this pool strategy the ``ResourceAwarePool``.
+
+Throughout the rest of this documentation page, we will use a "camp resource" to refer to a "stream of
+execution". If the camp resource is on the device, then we are referring to a device stream such 
+as a cuda stream or hip stream.
+
+Using a Single Resource
+-----------------------
+
+Umpire's strategies such as ``QuickPool`` and ``DynamicPoolList`` work very well
+on the device when we are dealing with a single camp device resource. In the figure below, we have
+the host resource which allocates memory (a\ :sub:`1`\), uses the memory in a kernel (k\ :sub:`1`\), then schedules
+a deallocate (d\ :sub:`2`\). Then, the host immidiately reuses that memory for a different kernel (k\ :sub:`2`\).
+
+.. image:: ./single_mem.png
+
+In this scenario, there is no potential for a data race, since we are dealing with just one cuda stream
+and kernels on a single stream execute sequentially. In other words, this scenario deals with only
+one Camp device resource. In this type of scenario, there is no need for a ``ResourceAwarePool`` because
+it would behave the same as your typical ``QuickPool``.
+
+.. note:: A ``ResourceAwarePool`` with only one resource will behave the same as a ``QuickPool`` strategy.
+   We don't advise using the ``ResourceAwarePool`` with only one resource since it will have the added
+   overhead of the pending state (explained below). Instead, just use ``QuickPool`` if possible. 
+
+Using Multiple Resources
+------------------------
+
+When dealing with multiple camp device resources, there is a possibility for a data race if we allocate, use, and 
+schedule a deallocation on one stream and then try to reuse that memory immediately on another stream. 
+The figure below depicts that scenario. Note that the overlap in the kernels corresponds to a potential
+data race.
+
+.. image:: ./multi-mem.png
+
+Umpire's ``ResourceAwarePool`` is designed to avoid any potential data races by making the resources
+"aware" of the memory used by another resource. If resource ``r2`` needs to allocate memory, but that
+memory is potentially still being used by another resource, ``r1``,  then ``r2``  will use different 
+memory instead. To do that, the ``ResourceAwarePool`` introduces a "pending" state. As soon as ``r1``
+schedules a deallocation, that memory is marked as ``_``pending``_``. Only once that pending memory has
```

Comment:
Wording suggestion: "As soon as ``r1`` schedules a deallocation, that memory is marked as ``_``pending``_`` and is only available for use by ``r1``.  When the deallocation is complete, the ``_``pending``_`` marker is cleared, making that memory available for use by other resources. So when ``r2`` needs an allocation, it first checks to see if the memory is still ``_``pending``_``.

Reply:
Done!